lib/spack/spack/test/cmd/buildcache.py

```diff
@@ -975,3 +976,259 @@ def mock_stat_url(url: str):
+        command_args = ["update-index", "--name", "test_view", "my-mirror", "DEADBEEF"]
+        buildcache(*command_args)
+
+    invalid_field_json = tmp_path / "invalid_field_json.json"
+    invalid_field_json.write_text('{"DEADBEEF": []}')
+    with pytest.raises(spack.error.SpackError, match=str(invalid_field_json)):
+        command_args = [
+            "update-index",
+            "--name",
+            "test_view",
+            "my-mirror",
+            str(invalid_field_json),
+        ]
+        buildcache(*command_args)
+
+
+def test_buildcache_create_view_empty(
+    tmp_path, mutable_config, mutable_database, mutable_mock_env_path
+):
+    mirror_directory = str(tmp_path)
+    mirror("add", "--unsigned", "my-mirror", mirror_directory)
+
+    # Push a spec to the cache
+    mpileaks_specs = mutable_database.query_local("mpileaks")
+    buildcache("push", "--update-index", "my-mirror", mpileaks_specs[0].format("{/hash}"))
+
+    # Make sure the view doesn't exist yet
+    with pytest.raises(spack.binary_distribution.FetchIndexError):
+        hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+
+    # Write a minimal lockfile (this is not validated/read by an enviornment)
+    empty_lockfile = tmp_path / "emptylock" / "spack.lock"
+    empty_lockfile.parent.mkdir()
+    empty_lockfile.write_text(
+        '{"_meta": {"lockfile-version": 1}, "roots": [], "concrete_specs": {}}', encoding="utf-8"
+    )
+    # Create a view with no specs
+    command_args = [
+        "update-index",
+        "--force",
+        "--name",
+        "test_view",
+        "my-mirror",
+        str(empty_lockfile),
+    ]
+    out = buildcache(*command_args)
+    assert "No specs found for view, creating an empty index" in out
+    hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+    assert not hashes_in_view
+
+
+@pytest.mark.parametrize("sources", ("hash", "specfile", "env", "lockfile"))
+def test_buildcache_create_view(
+    tmp_path, mutable_config, mutable_database, mutable_mock_env_path, sources
+):
+    mirror_directory = str(tmp_path)
+    mirror("add", "--unsigned", "my-mirror", mirror_directory)
+
+    # Push all of the mpileaks packages to the cache
+    mpileaks_specs = mutable_database.query_local("mpileaks")
+    for s in mpileaks_specs:
+        buildcache("push", "--update-index", "my-mirror", s.format("{/hash}"))
+
+    # Grab all of the hashes for each mpileaks spec
+    mpileaks_0_hashes = [s.dag_hash() for s in mpileaks_specs[0].traverse()]
+    mpileaks_1_hashes = [s.dag_hash() for s in mpileaks_specs[1].traverse()]
+
+    # Make sure the view doesn't exist yet
+    with pytest.raises(spack.binary_distribution.FetchIndexError):
+        hashes_in_view = read_specs_in_index(mirror_directory, "test_view")
+
+    # Create a view using a parametrized source that contains one of the mpileaks
+    with globals()[f"spec_{sources}"](mpileaks_specs[0], tmp_path / "source_stage") as source_args:
```

Comment:
The only place we use the `sources` string is in this lookup, can we change this to:

```suggestion
    with sources(mpileaks_specs[0], tmp_path / "source_stage") as source_args:
```
And then change the parametrization to
```
@pytest.mark.parametrize("sources", (spec_hash, spec_specfile, spec_env, spec_lockfile))
```
And the same for all of the analogous tests?

Reply:
Actually, I think we could do one better than that. We can make a single pytest fixture that inherits from `mutable_config`, `mutable_mock_env_path`, `tmpdir_factory`, and `requests`. Then it should be able to:

1. Only concretize the environment once
2. Remove the path as an argument (we only ever pass one value)
3. Check the `requests` fixture to know which type of source to return
4. Return a callable context manager that accepts a spec and does the appropriate thing in each case.

Reply:
I need to read this a couple more times to understand what you mean.

Reply:
I just went with the first option you gave here, easier for me to wrap my head around.